contracts/cw-names-registry/src/contract.rs

```diff
@@ -0,0 +1,256 @@
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let payment_token_address = deps.api.addr_validate(&msg.payment_token_address)?;
+    let admin = deps.api.addr_validate(&msg.admin)?;
+
+    if msg.payment_amount.is_zero() {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let config = Config {
+        admin,
+        payment_token_address,
+        payment_amount: msg.payment_amount,
+    };
+
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::new().add_attribute("action", "instantiate"))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(wrapped) => execute_receive(deps, env, info, wrapped),
+        ExecuteMsg::UpdateConfig {
+            new_payment_token_address,
+            new_admin,
+            new_payment_amount,
+        } => execute_update_config(
+            deps,
+            env,
+            info,
+            new_payment_token_address,
+            new_admin,
+            new_payment_amount,
+        ),
+        ExecuteMsg::Revoke { name } => execute_revoke(deps, env, info, name),
+        ExecuteMsg::Reserve { name } => execute_reserve(deps, env, info, name),
+        ExecuteMsg::TransferReservation { name, dao } => {
+            execute_transfer_reservation(deps, env, info, name, dao)
+        }
+    }
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // We only take payments from our specified token
+    if info.sender != config.payment_token_address {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // TODO: Is there a way we can verify this is a DAO
+    let sender = wrapped.sender;
+    let amount = wrapped.amount;
+    let msg: ReceiveMsg = from_binary(&wrapped.msg)?;
+
+    match msg {
+        ReceiveMsg::Register { name } => register_name(deps, env, sender, amount, name),
+    }
+}
+
+pub fn register_name(
+    deps: DepsMut,
+    _env: Env,
+    sender: String,
+    amount: Uint128,
+    name: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if amount < config.payment_amount {
+        return Err(ContractError::InsufficientFunds {});
+    }
+
+    // this is the DAO
+    let sender = deps.api.addr_validate(&sender)?;
+
+    if RESERVED_NAMES.has(deps.storage, name.clone()) {
+        return Err(ContractError::NameReserved {});
+    }
+
+    if NAME_TO_DAO.has(deps.storage, name.clone()) {
+        return Err(ContractError::NameAlreadyTaken {});
+    }
+
+    if DAO_TO_NAME.has(deps.storage, sender.clone()) {
+        return Err(ContractError::NameAlreadyRegistered {});
```

Comment:
See comment below about the name of this error. I actually missed that we were doing a one-name-per-DAO check here because I didn't grok the difference between `NameAlreadyTaken` and `NameAlreadyRegistered`.